core/id.go

```diff
@@ -0,0 +1,38 @@
+package core
+
+import (
+	"sync"
+)
+
+var mu sync.Mutex
+var turn, cycle, counter uint32 = 0, 0, 0
+
+var totalBits uint32 = 32
+var turnBits uint32 = 4
+var counterBits uint32 = totalBits - turnBits
+
+var cycleMap []uint32
+
+func init() {
+	cycleMap = make([]uint32, 1<<turnBits)
+}
+
+func ExpandID(id uint32) uint64 {
+	var _id uint64 = uint64(id)
+	_id |= uint64(cycleMap[id>>counterBits]) << counterBits
+	return _id
+}
+
+// TODO: Persisting the cycle on disk
+// and reloading it when we start the server
+func NextID() uint32 {
+	mu.Lock()
+	defer mu.Unlock()
+	counter = (counter + 1) & ((1 << counterBits) - 1)
+	if counter == 0 {
+		cycle++
+		turn = (turn + 1) & ((1 << turnBits) - 1)
```

Comment:
Q : If the turn reaches 0 again (wraps around after `turn=16`), it'll re-write the cycle. Is it expected that transaction IDs of earlier `turn=16` will no be present by the time this cycle comes through?

Reply:
Ah nvm, saw your buffer & disambiguate point, it's indeed not expected to be present.